crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,419 @@
+          # proposal from the queue and exit.
+          #
+          # remove from queue if it was queued before; might not be in the queue
+          # because the proposal got changed since it got added to the queue
+          unless preexisting_queued_item.nil?
+            logger.debug("queue proposal: dequeuing already queued #{inst} #{bc}")
+            dequeue_proposal_no_lock(bc, inst)
+          end
+
+          return [ delay, pre_cached_nodes ]
+        end
+
+        # Delay not empty, we're missing some nodes.
+        # And proposal is not in queue
+        if preexisting_queued_item.nil?
+          proposal_queue << { "barclamp" => bc, "inst" => inst, "elements" => elements, "deps" => deps }
+        else
+          # Update (overwrite) item that is already in queue
+          preexisting_queued_item["elements"] = elements
+          preexisting_queued_item["deps"] = deps
+          proposal_queue.save
+        end
+      rescue StandardError => e
+        logger.error("Error queuing proposal for #{bc}:#{inst}: #{e.message}")
+      ensure
+        file_lock.release(f)
+      end
+
+      # Mark the proposal as in the queue
+      prop = Proposal.where(barclamp: bc, name: inst).first
+      prop["deployment"][bc]["crowbar-queued"] = true
+      prop.save
+      logger.debug("queue proposal: exit #{inst} #{bc}")
+      [ delay, pre_cached_nodes ]
+    end
+
+    # Locking wrapper around dequeue_proposal_no_lock
+    def dequeue_proposal(bc, inst)
+      logger.debug("dequeue proposal: enter #{inst} #{bc}")
+      ret = false
+      begin
+        f = file_lock.acquire("queue", logger: logger)
+
+        if proposal_queue.empty?
+          logger.debug("dequeue proposal: exit #{inst} #{bc}: no entry")
+          return [200, {}]
+        end
+
+        dequeue_proposal_no_lock(bc, inst)
+      rescue StandardError => e
+        logger.error("Error dequeuing proposal for #{bc}:#{inst}: #{e.message} #{e.backtrace.join("\n")}")
+        logger.debug("dequeue proposal: exit #{inst} #{bc}: error")
+        return [400, e.message]
+      ensure
+        file_lock.release(f)
+      end
+      logger.debug("dequeue proposal: exit #{inst} #{bc}")
+      return dequeued ? [200, {}] : [400, '']
+    end
+
+    # Deps are satisfied if all exist, have been deployed and are not in the queue ATM.
+    def dependencies_satisfied?(deps)
+      item["deps"].all? do |dep|
+        depprop = Proposal.where(barclamp: dep["barclamp"], name: dep["inst"]).first
+        depprop_queued   = depprop["deployment"][dep["barclamp"]]["crowbar-queued"] rescue false
+        depprop_deployed = (depprop["deployment"][dep["barclamp"]]["crowbar-status"] == "success") rescue false
+
+        depprop && !depprop_queued && depprop_deployed
+      end
+    end
+
+    #
+    # NOTE: If dependencies don't form a DAG (Directed Acyclic Graph) then we have a problem
+    # with our dependency algorithm
+    #
+    def process_queue
```

Comment:
Metrics/AbcSize: Assignment Branch Condition size for process_queue is too high. [69.48/50]